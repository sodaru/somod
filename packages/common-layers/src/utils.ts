import { existsSync } from "fs";
import { mkdir, readdir, readFile, writeFile } from "fs/promises";
import { join } from "path";
import { chdir } from "process";
import { CommonLayers, layerLibraries } from "./index";
import rimraf from "rimraf";

const deleteDir = (dir: string): Promise<void> => {
  return new Promise((resolve, reject) => {
    rimraf(dir, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
};

/**
 * Generates layer structure and package.json(including dependencies) for all configured layers
 * layer structure will be like {base folder}/layers/{layer name}/nodejs/package.json
 */
export const generate = async () => {
  chdir(join(__dirname, ".."));
  await mkdir("./layers", { recursive: true });
  const rootPackageJsonStr = await readFile("./package.json", {
    encoding: "utf8"
  });

  const rootPackageJson = JSON.parse(rootPackageJsonStr);

  for (const layer in layerLibraries) {
    const dependencies = {};
    layerLibraries[layer]["libraries"].forEach((dependency: string) => {
      const dependencyVersion = rootPackageJson.peerDependencies[dependency];
      if (!dependencyVersion) {
        throw new Error(
          `Dependency ${dependency} in layer ${layer} is not found in root package.json's peerDependencies`
        );
      }
      dependencies[dependency] = dependencyVersion;
    });
    const packageJSON = {
      name: `slplayer-${layer.toLowerCase()}`,
      version: "1.0.0",
      description: "Function layer generated by @somod/common-layers",
      dependencies: dependencies
    };
    await mkdir(`./layers/${layer}/nodejs`, { recursive: true });
    await writeFile(
      `./layers/${layer}/nodejs/package.json`,
      JSON.stringify(packageJSON, null, 2)
    );
  }
};

/**
 * compares and removes duplicate libraries from passed layer w.r.t base layer
 * @param layerName layer name for comparing duplicate libraries
 */
export const removeBaseLayerLibraries = async (
  layerName: string
): Promise<void> => {
  chdir(join(__dirname, ".."));
  const baseLayerNodeModulesPath = `./layers/${CommonLayers.baseLayer}/nodejs/node_modules`;
  const layerNodeModulesPath = `./layers/${layerName}/nodejs/node_modules`;
  const dirs = await readdir(layerNodeModulesPath);
  const scopes: string[] = [];
  const libraries = dirs.filter(dir => {
    if (dir.startsWith("@")) {
      scopes.push(dir);
      return false;
    }
    return true;
  });

  await Promise.all(
    scopes.map(async scope => {
      const scopedLibraries = await readdir(`${layerNodeModulesPath}/${scope}`);
      libraries.push(...scopedLibraries.map(l => `${scope}/${l}`));
    })
  );

  await Promise.all(
    libraries.map(async lib => {
      if ([".bin", ".package-lock.json"].includes(lib)) {
        return;
      }
      if (existsSync(`${baseLayerNodeModulesPath}/${lib}`)) {
        const packageJsonStrList = await Promise.all([
          readFile(`${baseLayerNodeModulesPath}/${lib}/package.json`, {
            encoding: "utf8"
          }),
          readFile(`${layerNodeModulesPath}/${lib}/package.json`, {
            encoding: "utf8"
          })
        ]);
        const [baseLayerLibVersion, layerLibVersion] = packageJsonStrList.map(
          p => JSON.parse(p).version
        );
        if (baseLayerLibVersion == layerLibVersion) {
          await deleteDir(`${layerNodeModulesPath}/${lib}`);
        }
      }
    })
  );
};
