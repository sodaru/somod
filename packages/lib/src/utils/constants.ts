export const file_packageJson = "package.json";
export const file_packageLockJson = "package-lock.json";

export const file_dotenv = ".env";

export const file_tsConfigSomodJson = "tsconfig.somod.json";

export const file_nextConfigJs = "next.config.js";

export const key_somod = "somod";

export const key_type = "type";

export const key_module = "module";
export const key_typings = "typings";

export const key_sideEffects = "sideEffects";

export const key_main = "main";
export const key_jsnextMain = "jsnext:main";
export const key_files = "files";

export const key_serverlessSchema = "serverlessSchema";

export const key_somodPlugins = "somodPlugins";

export const path_nodeModules = "node_modules";

export const path_somodWorkingDir = ".somod";

export const path_build = "build";

export const path_ui = "ui";
export const path_lib = "lib";

export const path_pages = "pages";
export const path_pagesData = "pages-data";
export const path_public = "public";

export const file_index_js = "index.js";
export const file_index_dts = "index.d.ts";

export const path_git = ".git";

export const file_configYaml = "config.yaml";
export const file_configJson = "config.json";

export const file_extensionTs = "extension.ts";
export const file_extensionJs = "extension.js";

export const path_serverless = "serverless";
export const path_functions = "functions";
export const path_functionLayers = "functionLayers";
export const path_middlewares = "middlewares";
export const file_templateYaml = "template.yaml";
export const file_templateJson = "template.json";

export const file_samConfig = "samconfig.toml";

export const path_vercel = ".vercel";

export const cli_version_regex =
  /^([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+)?$/;

export const namespace_page = "UI Page";

export const namespace_public = "UI Public Asset";

export const namespace_env_config = "UI Env Config";

export const namespace_public_runtime_config = "UI Public Runtime Config";

export const namespace_server_runtime_config = "UI Server Runtime Config";

export const namespace_api_gateway = "Serverless Api Gateway";

export const namespace_parameter = "Parameter";

export const namespace_output = "Serverless Output";

export const resourceType_Function = "AWS::Serverless::Function";
export const resourceType_FunctionLayer = "AWS::Serverless::LayerVersion";
export const resourceType_FunctionMiddleware =
  "SOMOD::Serverless::FunctionMiddleware";

export const custom_resource_prefix = "Custom::";

export const file_parametersYaml = "parameters.yaml";

export const file_parametersJson = "parameters.json";

export const key_parameter_overrides = "parameter_overrides";
