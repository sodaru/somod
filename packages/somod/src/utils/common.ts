import { Command } from "nodejs-cli-runner";

export type SOMODCommandTypeOptions = {
  ui: boolean;
  serverless: boolean;
};

export type DebugModeOptions = {
  debug: boolean;
};

export const addSOMODCommandTypeOptions = (command: Command) => {
  command.option("--ui", "only ui");
  command.option("--serverless", "only serverless");
};

export const addDebugOptions = (command: Command) => {
  command.option("-d, --debug", "Enable Debug mode");
};

export const getSOMODCommandTypeOptions = (
  options: SOMODCommandTypeOptions
): SOMODCommandTypeOptions => {
  const ui = !options.serverless || options.ui;
  const serverless = !options.ui || options.serverless;
  return { ui, serverless };
};
