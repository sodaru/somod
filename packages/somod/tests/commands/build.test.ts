import buildCommand from "../../src/commands/build";
import { IContext } from "somod-types";
import * as lib from "somod-lib";
import { mockedFunction } from "../utils";
import { addCommonOptions } from "nodejs-cli-runner";

jest.mock("somod-lib", () => {
  const original = jest.requireActual("somod-lib");
  const mocked = {};
  for (const key in original) {
    mocked[key] =
      typeof original[key] === "function" ? jest.fn() : original[key];
  }
  return {
    __esModule: true,
    ...mocked
  };
});

describe("Test command build", () => {
  let output: unknown[][] = [];

  // eslint-disable-next-line no-console
  const originalConsoleLog = console.log;

  const functionsToBeMocked = [
    "buildParameters",
    "buildServerlessTemplate",
    "buildUiConfigYaml",
    "buildUiPublic",
    "compileTypeScript",
    "deleteBuildDir",
    "validatePageData",
    "validatePageExports",
    "isValidTsConfigSomodJson",
    "savePackageJson",
    "updateSodaruModuleKeyInPackageJson",
    "validatePackageJson",
    "validateParametersWithSchema",
    "validateServerlessTemplateWithSchema",
    "validateUiConfigYaml",
    "validateServerlessTemplate",
    "validateUiConfigYamlWithSchema",
    "validateFunctionExports",
    "initializeContext",
    "bundleExtension"
  ];

  const mock = (input: Record<string, unknown>) => {
    for (const key of functionsToBeMocked) {
      // eslint-disable-next-line import/namespace
      mockedFunction(lib[key]).mockImplementation(async (...args) => {
        output.push([key, ...args]);
        return input[key];
      });
    }
    mockedFunction(lib.findRootDir).mockImplementation((...args) => {
      output.push(["findRootDir", ...args]);
      return "/root/dir";
    });
  };

  beforeAll(() => {
    addCommonOptions(buildCommand);
  });

  beforeEach(() => {
    buildCommand.setOptionValue("verbose", undefined);
    buildCommand.setOptionValue("debug", undefined);
    buildCommand.setOptionValue("ui", undefined);
    buildCommand.setOptionValue("serverless", undefined);

    // eslint-disable-next-line no-console
    console.log = originalConsoleLog;
    for (const key of functionsToBeMocked) {
      // eslint-disable-next-line import/namespace
      mockedFunction(lib[key]).mockReset();
    }
    output = [];

    // eslint-disable-next-line no-console
    console.log = jest.fn();
  });

  test("without args", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          prebuildHooks: [],
          buildHooks: []
        }
      } as IContext
    });
    await buildCommand.parseAsync([], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with verbose", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          prebuildHooks: [],
          buildHooks: []
        }
      } as IContext
    });
    await buildCommand.parseAsync(["-v"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with debug mode", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          prebuildHooks: [],
          buildHooks: []
        }
      } as IContext
    });
    await buildCommand.parseAsync(["-d"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with serverless", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          prebuildHooks: [],
          buildHooks: []
        }
      } as IContext
    });
    await buildCommand.parseAsync(["--serverless"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with ui", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          prebuildHooks: [],
          buildHooks: []
        }
      } as IContext
    });
    await buildCommand.parseAsync(["--ui"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with all options", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          prebuildHooks: [],
          buildHooks: []
        }
      } as IContext
    });
    await buildCommand.parseAsync(["-v", "-d", "--ui", "--serverless"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with hooks from extensions", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          prebuildHooks: [
            {
              extension: "e1",
              value: context => {
                output.push(["pre-e1", context]);
              }
            },
            {
              extension: "e2",
              value: context => {
                output.push(["pre-e2", context]);
              }
            },
            {
              extension: "e3",
              value: context => {
                output.push(["pre-e3", context]);
              }
            }
          ],
          buildHooks: [
            {
              extension: "e1",
              value: context => {
                output.push(["e1", context]);
              }
            },
            {
              extension: "e2",
              value: context => {
                output.push(["e2", context]);
              }
            },
            {
              extension: "e3",
              value: context => {
                output.push(["e3", context]);
              }
            }
          ]
        }
      } as IContext
    });
    await buildCommand.parseAsync(["-v", "-d", "--ui", "--serverless"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });
});
