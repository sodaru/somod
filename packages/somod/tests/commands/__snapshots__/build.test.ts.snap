// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test command build with all options 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    true,
    true,
  ],
  Array [
    "validatePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "isValidTsConfigSomodJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateParametersWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYamlWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageData",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplateWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateFunctionExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "deleteBuildDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "compileTypeScript",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    true,
  ],
  Array [
    "bundleExtension",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    true,
  ],
  Array [
    "buildUiPublic",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "updateSodaruModuleKeyInPackageJson",
    "/root/dir",
  ],
  Array [
    "savePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command build with debug mode 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    true,
    true,
  ],
  Array [
    "validatePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "isValidTsConfigSomodJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateParametersWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYamlWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageData",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplateWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateFunctionExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "deleteBuildDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "compileTypeScript",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "bundleExtension",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "buildUiPublic",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "updateSodaruModuleKeyInPackageJson",
    "/root/dir",
  ],
  Array [
    "savePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command build with serverless 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    undefined,
    true,
    undefined,
  ],
  Array [
    "validatePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "isValidTsConfigSomodJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateParametersWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplateWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateFunctionExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "deleteBuildDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "compileTypeScript",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "bundleExtension",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "buildServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "updateSodaruModuleKeyInPackageJson",
    "/root/dir",
  ],
  Array [
    "savePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command build with ui 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    undefined,
    undefined,
  ],
  Array [
    "validatePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "isValidTsConfigSomodJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateParametersWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYamlWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageData",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "deleteBuildDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "compileTypeScript",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "bundleExtension",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "buildUiPublic",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "updateSodaruModuleKeyInPackageJson",
    "/root/dir",
  ],
  Array [
    "savePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command build with verbose 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    true,
    undefined,
  ],
  Array [
    "validatePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "isValidTsConfigSomodJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateParametersWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYamlWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageData",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplateWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateFunctionExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "deleteBuildDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "compileTypeScript",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    true,
  ],
  Array [
    "bundleExtension",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    true,
  ],
  Array [
    "buildUiPublic",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "updateSodaruModuleKeyInPackageJson",
    "/root/dir",
  ],
  Array [
    "savePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command build without args 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    true,
    undefined,
  ],
  Array [
    "validatePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "isValidTsConfigSomodJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateParametersWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYamlWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validatePageData",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplateWithSchema",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "validateFunctionExports",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "deleteBuildDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "compileTypeScript",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "bundleExtension",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "buildUiPublic",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildUiConfigYaml",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildServerlessTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "buildParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
  Array [
    "updateSodaruModuleKeyInPackageJson",
    "/root/dir",
  ],
  Array [
    "savePackageJson",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "buildHooks": Array [],
        "prebuildHooks": Array [],
      },
    },
  ],
]
`;
